From bb9b0383745fa24c9cac16451428cca00c15350b Mon Sep 17 00:00:00 2001
From: Mario Schuknecht <mario.schuknecht@dresearch-fe.de>
Date: Wed, 12 Jun 2013 15:15:25 +0200
Subject: [PATCH] hikirk: get mac address as command line parameter

The two hikirk network interfaces get their MAC addresses from the kernel
command line. "ethaddr" corresponds to eth0 and "eth1addr" corresponds to
eth1.

Signed-off-by: Mario Schuknecht <mario.schuknecht@dresearch-fe.de>
---
 arch/arm/mach-kirkwood/openrd-setup.c |   33 +++++++++++++++++++++++++++++++++
 1 file changed, 33 insertions(+)

diff --git a/arch/arm/mach-kirkwood/openrd-setup.c b/arch/arm/mach-kirkwood/openrd-setup.c
index 1adb77b..a7235c5 100644
--- a/arch/arm/mach-kirkwood/openrd-setup.c
+++ b/arch/arm/mach-kirkwood/openrd-setup.c
@@ -160,6 +160,39 @@ static struct platform_device openrd_client_audio_device = {
 	.id		= -1,
 };
 
+static void __init set_ethaddr(char *str, struct mv643xx_eth_platform_data* eth_data)
+{
+	int i;
+	unsigned int mac[6];
+
+	if(!str) { return; }
+
+	if(6 == sscanf(str, "%x:%x:%x:%x:%x:%x", &mac[0], &mac[1], &mac[2], 
+						&mac[3], &mac[4], &mac[5])) {
+		for(i=0; i<6; ++i) {
+			eth_data->mac_addr[i] = (u8)(mac[i] & 0xFF);
+		}
+	} else {
+		printk(KERN_WARNING "Command line mac invalid: '%s'\n", str);
+	}
+}
+
+static int __init set_eth0addr(char *str)
+{
+	set_ethaddr(str, &openrd_ge00_data);
+	return 1;
+}
+
+static int __init set_eth1addr(char *str)
+{
+	set_ethaddr(str, &openrd_ge01_data);
+	return 1;
+}
+
+/* Parse boot_command_line string ethaddr */
+__setup("ethaddr=", set_eth0addr);
+__setup("eth1addr=", set_eth1addr);
+
 static int __initdata uart1;
 
 static int __init sd_uart_selection(char *str)
-- 
1.7.9.5

